stdlib/README.adoc

```diff
@@ -0,0 +1,208 @@
+as / already superseded by external projects.
+
+On the other hand, various features would be considered by many as
+being of general interest in 2018, at different degrees, and might
+thus deserve some lightweight support in the stdlib: some (limited)
+support for Unicode (e.g. Uchar, perhaps utf8<->latin1 conversion),
+simple encoding/escaping stuff (e.g. Base64), a few more
+datastructures (e.g. bitsets), perhaps even a definition of a JSON
+datatype/parser/printer. Opinions will necessarily vary here. Who will
+decide, and based on which criteria?
+
+== Criteria for extending stdlib
+
+The size in itself should not really be a direct criterion for
+accepting or rejecting a proposed addition to the stdlib, I'd say. We
+should rather focus on other criteria (which are affected by the size,
+but not only):
+
+- The efforts required to maintain it (taking backward compatibility
+  into account).
+- The code and API quality (including the documentation).
+- How general purpose the addition is.
+- The usefulness of extending the stdlib vs maintaining the addition
+  as an external project.
+- In particular, the risk than an external project can do something
+  clearly better (which would leave us in the uncomfortable situation
+  where we need to keep something in the stdlib for backward
+  compatibility while knowing it isn't a good solution to recommend).
+
+For instance, adding a small function to an existing opaque
+datastructure (Set, Map, Stack, Buffer) would generally induce a very
+small maintenance cost and it's not really possible to do that outside
+the stdlib (short of completely reimplementing the datastructure
+altogether). The argument is weaker for extending e.g. List or Array,
+since this can more easily be done outside stdlib. That said, if the
+addition is useful enough, there is no strong reason not to accept it.
+
+
+== More to discuss
+
+* When a module `M` has a main type it must be named `M.t`.
+* Functions should take the main type `t` as first argument (unless
+  there is a good reason for it, see next point).
+* If a functions takes another function as argument, that argument
+  should have the label `f` and not be last (to be able to omit the
+  label).
+* When a function may raise an exception as a non-exceptional
+  behavior, its name must have the suffix `_exn` to remind the user
+  that the exception should be handled.  Another function returning an
+  option instead must be provided.  Example:
+[source,ocaml]
+----
+val find : 'a t -> key -> 'a option
+val find_exn : 'a t -> key -> 'a
+----
+
+
+=== Collections
+
+* If a module `M` defines a collection of values, say `'a M.t`, it
+  must also define the following standard higher order functions:
+
+[source,ocaml]
+----
+val iter : f:('a -> unit) -> 'a t -> unit
+val iteri : f:(int -> 'a -> unit) -> 'a t -> unit
+val map : f:('a -> 'b) -> 'a t -> 'b t
+val mapi : f:(int -> 'a -> 'b) -> 'a t -> 'b t
+val fold : f:('a -> 'b -> 'a) -> init:'a -> 'b t -> 'a
```

Comment:
I'm not sure `iteri` and `mapi` make sense for all collections.  They do make sense for lists and arrays and things that have an `nth` operation, but it's less clear to me we want them for, say, sets or maps or hashtables.

Also: what about supporting the `Seq` interface?  That's something we could expect from all collections.

Reply:
> what about supporting the `Seq` interface?

This is [part of the proposal](https://github.com/ocaml/ocaml/pull/2111/files#diff-a4ef3647120d442534b39401cf66d465R145).

Reply:
Good! Sorry I missed it.